Update contribution guide (#1121)
mattermost/mattermost-developer-documentation: site/content/contribute/why-contribute/_index.md

---
title: "Why you might contribute"
heading: "Why you might contribute"
weight: 1
---

> ## Subsection 1 — You’ve found a bug

Reviewer: ```suggestion
## You’ve found a bug
```

1. If the bug fix that you’re proposing would be larger than 20 lines of code, [create a GitHub issue](https://github.com/mattermost/mattermost-server/issues).

>     1. You can speed up the process by asking about the issue in the  [Contributors](https://community.mattermost.com/core/channels/tickets) or  [Developers](https://community.mattermost.com/core/channels/developers) channels.

Reviewer: ```suggestion
    - You can speed up the process by asking about the issue in the  [Contributors](https://community.mattermost.com/core/channels/tickets) or  [Developers](https://community.mattermost.com/core/channels/developers) channels on the Mattermost Community Server.
```

>     2. [Here’s a good example of a contribution](https://github.com/mattermost/mattermost-mobile/pull/364/commits/7a97451b62fee4022edac4c0395ad0a5cbf1bb66) that is small enough to not need a ticket while still being incredibly helpful.

Reviewer: ```suggestion
    - [Here’s a good example of a contribution](https://github.com/mattermost/mattermost-mobile/pull/364/commits/7a97451b62fee4022edac4c0395ad0a5cbf1bb66) that is small enough to not need a ticket while still being incredibly helpful.
```

> 2. If you’ve volunteered to take the ticket once its active, or if your fix is too small to warrant the ticket, fork the applicable repo and start making your changes.

Reviewer: ```suggestion
2. If you’ve volunteered to take the ticket once it's active, or if your fix is too small to warrant the ticket, fork the applicable repository, and start making your changes.
```

> 3. Create a PR on the `master` branch of the applicable repo. If there is an associated [Jira](https://mattermost.atlassian.net/issues/?jql=) or GitHub issue, your PR should begin with the issue ID (e.g.  `[MM-394]` or  `[GH-394]`). Our GitHub PR template will walk you through the creation of your PR.

Reviewer: ```suggestion
3. Create a PR on the `master` branch of the applicable repository. If there is an associated [Jira](https://mattermost.atlassian.net/issues/?jql=) or GitHub issue, your PR should begin with the issue ID (e.g.  `[MM-394]` or  `[GH-394]`). Our GitHub PR template will walk you through the creation of your PR.
```

> 4. If you’re a community contributor, the team at Mattermost will handle most of the review process for you. You can wait for a reviewer to be assigned and for them to review your work. Generally the process works like you’d expect: they’ll make suggestions, you’ll implement them and discuss it in the PR comments. If you’re having trouble getting a reviewer to be assigned to your PR, take a look at GitHub’s suggested reviewers and reach out to them in the [Community workspace](http://community.mattermost.com).

Reviewer: ```suggestion
4. If you’re a community contributor, the team at Mattermost will handle most of the review process for you. You can wait for a reviewer to be assigned and for them to review your work. Generally the process works like you’d expect: they’ll make suggestions, you’ll implement them, and together you'll discuss changes in PR comments. If you’re having trouble getting a reviewer to be assigned to your PR, take a look at GitHub’s suggested reviewers and reach out to them in the [Community workspace](http://community.mattermost.com).
```

5. Here is an overview of how our review process works. If you’re a core contributor, you can manage the process yourself.

>     1. Start by adding the  `1: UX Review`,  `2: Dev Review`, and  `3: QA Review`  labels, as well as whichever are applicable from [the list of labels]({{< ref "/contribute/more-info/getting-started/labels" >}}).

Reviewer: ```suggestion
    a. Start by adding the  `1: UX Review`,  `2: Dev Review`, and  `3: QA Review`  labels, as well as whichever are applicable from [the list of labels]({{< ref "/contribute/more-info/getting-started/labels" >}}).
```

>     2. Add a milestone or a cherry-picking label if the PR needs it. Most don’t and will simply ship once merged, but if there are upcoming milestones, some reviewers are going to prioritize reviews attached to those milestones. Note that adding a milestone is mandatory for bug fixes that must be cherry-picked.

Reviewer: ```suggestion
    b. Add a milestone or a cherry-picking label if the PR needs it. Most don’t and will simply ship once merged, but if there are upcoming milestones, some reviewers are going to prioritize reviews attached to those milestones. Note that adding a milestone is mandatory for bug fixes that must be cherry-picked.
```

>     3. Likely you know whose code you’re changing, so assign the appropriate PM or Designer first for a UX review. When in doubt, somebody is better than anybody, as they can always reassign. Once they give the OK, remove the `1: UX Review` label.

Reviewer: ```suggestion
    c. Likely you know whose code you’re changing, so assign the appropriate PM or Designer first for a UX review. When in doubt, somebody is better than anybody, as they can always reassign. Once they give the OK, remove the `1: UX Review` label.
```

>     4. Request two core committers to review — one with experience solving that particular or in that particular language, and another with expertise in the domain of your changes. Don’t request the same people over and over again unless they’re the only people with the necessary expertise. Once they give the OK, remove the `2: Dev Review` label.

Reviewer: ```suggestion
    d. Request two core committers to review — one with experience solving that particular or in that particular language, and another with expertise in the domain of your changes. Don’t request the same people over and over again unless they’re the only people with the necessary expertise. Once they give the OK, remove the `2: Dev Review` label.
```

>     5. Request a QA tester from the team that owns the code that you’re changing. They’ll decide the necessary scope of testing. Make sure you’ve defined what the expected results are so they know when their tests are successful. Once they give the OK, remove the `3: QA Review` label. The QA tester may add the `QA Review Done` label as well.

Reviewer: ```suggestion
    e. Request a QA tester from the team that owns the code that you’re changing. They’ll decide the necessary scope of testing. Make sure you’ve defined what the expected results are so they know when their tests are successful. Once they give the OK, remove the `3: QA Review` label. The QA tester may add the `QA Review Done` label as well.
```

>         1. Generally avoid doing QA testing yourself, unless it’s obvious that it’s not necessary. In any case, it likely won’t be obvious to onlookers, so clearly explain the rationale in the comments on the PR.

Reviewer: ```suggestion
        **Tip:** We recommend that you generally avoid doing QA testing yourself, unless it’s obvious that it’s not necessary. The best way to help Mattermost QA is to clearly explain your changes and expected outcomes in the PR.
```

>     6. Once all outstanding requests are completed and all involved have approved of the PR, add the `4: Reviews Complete` if it hasn’t been already, merge the PR into `master`, and delete the old branch. If your pull request depends on other pull requests, consider assigning the  `Do Not Merge/Awaiting PR`  label to avoid merging prematurely.

Reviewer: ```suggestion
    f. Once all outstanding requests are completed and all involved have approved of the PR, add the `4: Reviews Complete` if it hasn’t been already, merge the PR into `master`, and delete the old branch. If your pull request depends on other pull requests, consider assigning the  `Do Not Merge/Awaiting PR`  label to avoid merging prematurely.
```

>     7. If this is a cherry-picked PR, the automation should handle everything from here on out. If the automated cherry-pick fails, you will need to cherry-pick the PR manually. Cherry-pick the master commit back to the appropriate releases. If the release branches have not been cut yet, leave the labels as-is and cherry-pick once the branch has been cut. The release manager will remind you to finish your cherry-pick. Set the  `CherryPick/Done`  label when you’re done with this process.

Reviewer: ```suggestion
    g. If this is a cherry-picked PR, the automation should handle everything from here on out. If the automated cherry-pick fails, you will need to cherry-pick the PR manually. Cherry-pick the master commit back to the appropriate releases. If the release branches have not been cut yet, leave the labels as-is and cherry-pick once the branch has been cut. The release manager will remind you to finish your cherry-pick. Set the  `CherryPick/Done`  label when you’re done with this process.
```

>     8. Update the appropriate Jira ticket so everybody knows where the project stands. Resolve the ticket for QA from “Ready for QA” button with QA test steps, or with “No Testing Required” if no QA testing is needed.

Reviewer: ```suggestion
    h. Update the appropriate Jira ticket so everybody knows where the project stands. Resolve the ticket for QA from “Ready for QA” button with QA test steps, or with “No Testing Required” if no QA testing is needed.
```

>     9. If you need to test your changes on a test server, you can add the appropriate label to the PR and wait 3-5 minutes for the server to be created (see [the labels page]({{< ref "/contribute/more-info/getting-started/labels.md)" >}}). A bot will send you your link and credentials in the form of a comment on the PR. Remove the label when you’re done to destroy the test server.

Reviewer: ```suggestion
    i. If you need to test your changes on a test server, you can add the appropriate label to the PR and wait 3-5 minutes for the server to be created (see [the labels page]({{< ref "/contribute/more-info/getting-started/labels.md)" >}}). A bot will send you your link and credentials in the form of a comment on the PR. Remove the label when you’re done to destroy the test server.
```

6. Once you address suggestions a reviewer has made, re-request a review from them. Their initial review was technically completed, so it’s no longer in their queue until you re-request.

>     1. Give reviewers time — they may have long queues and different schedules. If you’ve been assigned a reviewer but haven’t heard from them in 5 business days, politely mention them in a PR comment.

Reviewer: ```suggestion
    **Tip:** Give reviewers time — they may have long queues and different schedules. If you’ve been assigned a reviewer but haven’t heard from them in five business days, you can politely bring focus back to your PR by mentioning them in a PR comment.
```

7. Mattermost has a system to categorize the inactivity of contributions. Invalid PRs don’t need to go through this cycle at the Community Coordinator’s discretion.

>     1. After 10 days of inactivity, a contribution becomes stale, and a bot will add the  `lifecycle/1:stale`  label to the PR. It’s on the Community Coordinator to nudge the right people to get a stale PR active again, even if that means clarifying requests so the contributor has more information to go off of.

Reviewer: ```suggestion
    - After ten days of inactivity, a contribution becomes stale, and a bot will add the  `lifecycle/1:stale`  label to the PR. It’s the job of the Community Coordinator to nudge the right people to get a stale PR active again, even if that means clarifying requests so the contributor has more information to work with.
```

    2. After 20 days of inactivity, a contribution becomes inactive, and a bot will add the `lifecycle/2:inactive` label to the PR. The Community Coordinator warns everybody involved how much time they have before the contribution is closed and again tries to reach out to the blocking party to help. They’re also going to make sure that it’s not the reviewers taking the PR to this point — contributions should only ever be inactive because of no response from the contributor.

>         1. When contributions are inactive but for good reason (for example, when the team is actively discussing a major design decision but they haven’t decided on anything yet), `lifecycle/frozen` would be a better label.

Reviewer: ```suggestion
        When contributions are inactive, but there's a good reason (for example, when the team is actively discussing a major design decision but they haven’t decided on anything yet), `lifecycle/frozen` would be a better label.
```

>         2. Inactive contributions are eligible to be picked up by another community member.

Reviewer: ```suggestion
        Inactive contributions are eligible to be picked up by another community member.
```

>     3. After 30 days of inactivity, a contribution becomes orphaned, and they’ll add the `lifecycle/3:orphaned` label to the now-closed PR. The associated `Help Wanted` ticket is given back its `Up For Grabs` status so others can pick up on the issue.

Reviewer: ```suggestion
    - After 30 days of inactivity, a contribution becomes orphaned, the `lifecycle/3:orphaned` label is added to the now-closed PR. The associated `Help Wanted` ticket is given back its `Up For Grabs` status so others can pick up the issue.
```

> ## Subsection 2 — You want to help with content

Reviewer: ```suggestion
## You want to help with content
```

    5. The best place to discuss problems with us is in the [Documentation Working Group channel](https://community.mattermost.com/core/channels/dwg-documentation-working-group) where you can ping our technical writers with the group `@docsteam`.
3. If you’d like to contribute to our blog, website, or social media, you also have a few options:
    1. You can get paid to write technical content for developers through the Mattermost [community writing program](https://mattermost.com/blog/blog-announcing-community-writing-program/).
    2. If you see a problem with any webpages, blog posts, or other content on [Mattermost.com](https://mattermost.com), you can notify us via [the Content channel](https://community.mattermost.com/core/channels/mattermost-blog).
    3. Share your contributor or user experience! Mention us when you promote work you do in our community and we’ll amplify the message through Mattermost social media profiles.
    4. Want to lead a social community? We can provide advice and resources to help you in the [Community channel](https://community.mattermost.com/core/channels/community-team).

> ## Subsection 3 — You’d like to make something more inclusive or accessible

Reviewer: ```suggestion
## You’d like to make something more inclusive or accessible
```

1. Accessibility is one of the most overlooked yet most important features of modern software development, and we’re eager to improve on the accessibility of the features in our open source project and its documentation.
2. Problems with the accessibility or the inclusivity of both features in the codebase and pieces of content describing Mattermost are bugs, so we treat them as such in our development process. See subsections 1 and 2 for the description on how you can submit your contribution.
3. When you contribute a change that incorporates an adjustment based on the principles of accessibility and inclusivity, please use this guide to back you up in the PR, ticket, or post.

> ## Subsection 4 — You have a feature idea

Reviewer: ```suggestion
## You have a feature idea
```

    Generally the process works like you’d expect: they’ll make suggestions, you’ll implement them, and together you'll discuss changes in PR comments. If you’re having trouble getting a reviewer to be assigned to your PR, take a look at GitHub’s suggested reviewers and reach out to them in the [Community workspace](http://community.mattermost.com).
5. Here is an overview of how our review process works listed in the order in which things happen. If you’re a core contributor, you can manage the process yourself.

    - **Labels will be added to your PR**, such as  `1: UX Review`,  `2: Dev Review`, and  `3: QA Review`, as applicable. See [the list of labels]({{< ref "/contribute/more-info/getting-started/labels" >}}) for details.
    
    - **A milestone or a `cherry-pick` label may be added to your PR**. Many PRs don’t need this step, and will simply ship once merged. However, if there are upcoming milestones, some reviewers are going to prioritize reviews attached to those milestones. Adding a milestone is mandatory for bug fixes that must be cherry-picked.
    

>     - In many cases, you'll know whose code you’re changing and can **assign the appropriate Product Manager, Designer, or Technical Writer to review your PR** from a UI/UX perspective. When in doubt, somebody is better than nobody, as reviewers can always reassign. Once reviwers approve your PR, the the `1: UX Review` label will be removed.

Reviewer 2: ```suggestion
    - In many cases, you'll know whose code you’re changing and can **assign the appropriate Product Manager, Designer, or Technical Writer to review your PR** from a UI/UX perspective. When in doubt, somebody is better than nobody, as reviewers can always reassign. Once reviewers approve your PR, the `1: UX Review` label will be removed.
```

      - When contributions are inactive, but there's a good reason (for example, when the team is actively discussing a major design decision but they haven’t decided on anything yet), `lifecycle/frozen` would be a better label.
      - Inactive contributions are eligible to be picked up by another community member.
    - **After 30 days of inactivity**: A contribution becomes orphaned, the `lifecycle/3:orphaned` label is added to the now-closed PR. The associated `Help Wanted` ticket is given back its `Up For Grabs` status so others can pick up the issue.

## You want to help with content

Good product and developer documentation content is as important as good code! If you notice and fix a content error in the documentation, in a repository README, or in another open source article describing Mattermost, we consider you to be as valued a member of our contributor community as those who contribute to core code.

1. If you see a problem with Mattermost [developer](https://developers.mattermost.com/) or [product](https://docs.mattermost.com/) documentation, you have a few options:

>     - If you have time to fix the mistake and it only affects a single page, navigate to the applicable page and select **Edit in GitHub** at the top right. You'll be walked through the process of creating a fork so that you can then follow the steps under the section titled “You’ve found a bug” in this guide.

Reviewer 2: ```suggestion
    - If you have time to fix the mistake and it only affects a single page, navigate to the applicable page and select **Edit in GitHub** at the top right. You'll be walked through the process of creating a fork so that you can then follow the steps under the section titled [“You’ve found a bug”]({{< ref "/contribute/why-contribute/#youve-found-a-bug" >}}) in this guide.
```

    - If you don’t have time to fix the mistake, copy the file path you’re on, and create a GitHub issue about the problem you found on the applicable repository. Make sure to include the file path and fill out the issue template completely to maximize clarity.
    - If you’re not up for creating a GitHub issue right now, that’s alright too! In the bottom-right corner of every product documentation page is the question “Did you find what you were looking for?” Use this to quickly provide direct feedback about any page you’re viewing.

>     - If you want to fix a larger problem that affects multiple pages or the structure of the docs, you should first report it as an issue on the appropriate GitHub repository, and follow the steps under  “You’ve found a bug”. The developer and product documentation repositories contain instructions on how to build and modify the sites locally so you can test larger changes more efficiently.

Reviewer 2: ```suggestion
    - If you want to fix a larger problem that affects multiple pages or the structure of the docs, you should first report it as an issue on the appropriate GitHub repository, and follow the steps under  [“You’ve found a bug”]({{< ref "/contribute/why-contribute/#youve-found-a-bug" >}}). The developer and product documentation repositories contain instructions on how to build and modify the sites locally so you can test larger changes more efficiently.
```

    -  You can get paid to write technical content for developers through the Mattermost [community writing program](https://mattermost.com/blog/blog-announcing-community-writing-program/).
    - If you see a problem with any webpages, blog posts, or other content on [Mattermost.com](https://mattermost.com), you can notify us via [the Content channel](https://community.mattermost.com/core/channels/mattermost-blog) on the Mattermost Community Server.
    - Share your contributor or user experience! Mention us when you promote your work within our community, and we’ll amplify the message through Mattermost social media platforms.
    - Want to lead a social community? We can provide advice and resources to help you in the [Community channel](https://community.mattermost.com/core/channels/community-team) on the Mattermost Community Server.

## You want to make something more inclusive or accessible

Accessibility is one of the most overlooked yet most important features of modern software development, and we’re eager to improve on the accessibility of the features in our open source project and its documentation.

> - Problems with the accessibility or the inclusivity of both features in the codebase and pieces of content describing Mattermost are bugs, so we treat them as such in our development process. 

Reviewer 2: ```suggestion
- Problems with the accessibility or the inclusivity of both features in the codebase, and pieces of content describing Mattermost, are bugs, so we treat them as such in our development process. 
```

> - When you contribute a change that incorporates an adjustment based on the principles of accessibility and inclusivity, please use this guide to back you up in the PR, ticket, or post.

Reviewer 2: "please use this guide"
we should link to the guide - if there is none yet, we can mark it as a todo

Reviewer: I believe we're referring to this new contribution guide on the dev docs site here.

Reviewer 2: ooooh lol
```suggestion
- When you contribute a change that incorporates an adjustment based on the principles of accessibility and inclusivity, please circle back to this guide to back you up in the PR, ticket, or post.
```

Each localization community is going to have specific guidelines on how to maintain Mattermost’s distinctive voice across language barriers. Read these guides thoroughly before starting to translate:

- [German](https://gist.github.com/der-test/6e04bff8a173e053811cb93e08838ca2)
- [French](https://github.com/wget/mattermost-localization-french-translation-rules)
- [Dutch](https://github.com/ctlaltdieliet/mattermost-localization-dutch-translation-rules)

To get started:

1. Join [the Mattermost translation server](http://translate.mattermost.com/) and [the localization community channel](https://community.mattermost.com/core/channels/localization) on the Mattermost Community Server.

   - If your preferred language already exists on the translation server, you can start making translations or translation suggestions immediately *there on the translation server*. Don’t try to do this manually through GitHub.

   - If your language is absent from the translation server, ask on the localization community channel for it to be added.
2. Each language in whole is assigned one quality level, and with each release cycle, it can be upgraded or downgraded if necessary.

>     - **Official** — 100% of the language’s translations are verified both by a Mattermost functionality expert and by a expert speaker of the target language. The language has at least one official reviewer who updates all of the strings that have changed in the English source text, and they have successfully kept all of the translated strings updated since the last release cycle. The language has been in use for at least three full release cycles.

Reviewer 2: Small change - sorry lol
But does it make sense to make the description into bullet points or line breaks for readability?
```suggestion
    - **Official** — 100% of the language’s translations are verified both by a Mattermost functionality expert and by an expert speaker of the target language. The language has at least one official reviewer who updates all of the strings that have changed in the English source text, and they have successfully kept all of the translated strings updated since the last release cycle. The language has been in use for at least three full release cycles.
```

>     - **Beta** — At least 90% of the language’s translations are verified by at least one Mattermost functionality expert who is fluent in the target language. If a language is at risk for ongoing maintenance, _____ can raise the threshold closer to 100%. Up to 10% of the translations may not be synced with the latest English source text.

Reviewer: ```suggestion
    - **Beta** — At least 90% of the language’s translations are verified by at least one Mattermost functionality expert who is fluent in the target language. If a language is at risk for ongoing maintenance, Mattermost can raise the threshold closer to 100%. Up to 10% of the translations may not be synchronized with the latest English source text.
```

    2. If your language is absent from the translation server, ask on the localization community channel for it to be added.
4. Each language in whole is assigned one quality level, and with each release cycle, it can be upgraded or downgraded if necessary.
    1. Official — 100% of the language’s translations are verified both by a Mattermost functionality expert and by a expert speaker of the target language. The language has at least one official reviewer who updates all of the strings that have changed in the English source text, and they have successfully kept all of the translated strings updated since the last release cycle. The language has been in use for at least three full release cycles.
    2. Beta — At least 90% of the language’s translations are verified by at least one Mattermost functionality expert who is fluent in the target language. If a language is at risk for ongoing maintenance, _____ can raise the threshold closer to 100%. Up to 10% of the translations may not be synced with the latest English source text.
    3. Alpha — The language does not meet the qualifications for the Beta level.

> 5. Every Monday, PRs gets opened for all updates. These PRs will be checked for unexpected character insertions and security problems. Reviewers should always “merge commit” or “rebase and merge” into the PR, but never, ever “squash and commit”. at 22:00 GMT that day, the PRs are approved and merged into the proper repos.

Reviewer: Good point. I'll review and update where needed. I'm inclined to remove the specific timeframe details for now.
